documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,296 @@
+        /// </summary> 
+        /// <returns>
+        /// A Stream object to read
+        /// </returns> 
+        public Stream CollectTracing1(SessionConfiguration configuration)
+
+        /// <summary>
+        /// Start tracing the application via CollectTracing2 command.
+        /// </summary> 
+        /// <returns>
+        /// A Stream object to read
+        /// </returns> 
+        public Stream CollectTracing2()
+
+        /// <summary>
+        /// Stop tracing the applicating via the StopTracing command.
+        /// </summary> 
+        /// <returns>
+        /// true if stopping the tracing session succeeded. false otherwise. 
+        /// </returns> 
+        public bool StopTracing()
+    }
+}
+```
+
+### DumpCommandHandler
+This is a CommandHandler class for sending Dump commands across the diagnostics IPC channel.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class DumpCommandHandler
+    {
+        public DumpCommandHandler(int processId);
+        public int GenerateCoreDump();
+    }
+}
+```
+
+
+### ProfilerCommandHandler
+This is a CommandHandler class for sending Profiler commands across the diagnostics IPC channel.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class ProfilerCommandHandler
+    {
+        public ProfilerCommandHandler(int processId);
+        public int AttachProfiler(uint attachTimeout, Guid profilerGuid, string profilerPath, byte[] additionalData);
+    }
+}
+```
+
+
+### Exceptions that can be thrown 
+```cs
+namespace Microsoft.Diagnostics.Client.Exceptions
+{
+    public class UnknownCommandException : Exception {}
+    public class UnknownMagicException : Exception {}
+    public class EventPipeInvalidArgumentException : Exception {}
+    public class DumpInvalidArgumentException : Exception {}
+    public class ProfilerInvalidArgumentException : Exception {}
+}
+```
+
+## Micorosft.Diagnostics.Client.EventPipe namespace
+
+### EventPipeCommandSet
+An enum for all EventPipe commands.
+``` cs
+namespace Micorosft.Diagnostics.Client.EventPipe
+{
+    public enum EventPipeCommandSet
+    {
+        StopTracing,
+        CollectTracing,
+        CollectTracing2
+    }
+}
+```
+
+### EventPipeSerializationFormat
+An enum for all EventPipe serialization format.
+```cs
+namespace Microsoft.Diagnostics.Client.EventPipe
+{
+    public enum EventPipeSerializationFormat
+    {
+        NetPerf,
+        NetTrace
+    }
+}
+```
+
+### EventPipeProvider
+A class that describes an EventPipe provider.
+```cs
+namespace Microsoft.Diagnostics.Client.EventPipe
+{
+    public class EventPipeProvider
+    {
+        public EventPipeProvider(
+            string name,
+            ulong keywords = ulong.MaxValue,
+            EventLevel eventLevel = EventLevel.Verbose,
+            string filterData = null)
+
+        public ulong Keywords { get; }
+
+        public EventLevel EventLevel { get; }
+
+        public string Name { get; }
+
+        public string FilterData { get; }
```

Comment:
I think we should make this an `IEnumerable<KeyValuePair<string, string>>` since that more closely maps to what the `EventSource` APIs are expecting.  The IPC protocol takes it as a single string, but escaping the input with quotes and escaping the quotes has proved confusing for users of `dotnet-trace` to say the least.  If we change the inputs to to the library, we can make the escaping and quoting an implementation detail rather than a concern of the end user.